_posts/2019-02-20-refactoring-state-pattern.markdown

```diff
@@ -0,0 +1,114 @@
+---
+layout: post
+title:  "Refactoring with Design Patterns - The State Pattern"
+date: 2019-02-20 10:00:00
+categories: ["code-refactor", "design-patterns"]
+author: "cleiviane"
+---
+
+In this series of [code refactoring](https://www.ombulabs.com/blog/tags/code-refactor) posts we are discussing how design patterns can be used to make our [Ruby](https://www.ruby-lang.org/en/) code beautiful, maintainable and clean.
+
+Today I want to talk about a pattern that can be very useful when we need to control the flow of a set of events of our objects: *The State Pattern* a.k.a *Finite State Machine*.
+
+As a developer it is common to see objects changing of states. At the beginning managing the state of an object can be as simple as having some boolean attributes where you can check if the object is in state A or B. But when the complexity increases you can end up with a number of states that are difficult to manage without breaking the [SOLID principles](http://butunclebob.com/ArticleS.UncleBob.PrinciplesOfOod). That is where we can benefit from the elegant solution the State Pattern gives to us.
```

Comment:
@cleicar This sounds weird: 

As a developer it is common to see objects changing of states.

Maybe you mean: 

As a developer it is common to see objects changing their state.

Reply:
@cleicar This sounds weird:

That is where we can benefit from the elegant solution the State Pattern gives to us.

Maybe it would sound better:

That is where we can implement the elegant solution provided by the State Pattern.